framework/src/Volo.Abp.AspNetCore/Microsoft/Extensions/DependencyInjection/AbpAspNetCoreServiceCollectionExtensions.cs

```diff
@@ -1,11 +1,21 @@
-﻿using Microsoft.AspNetCore.Hosting;
+﻿using System.Linq;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.Extensions.Hosting;
 
 namespace Microsoft.Extensions.DependencyInjection
 {
     public static class AbpAspNetCoreServiceCollectionExtensions
     {
         public static IWebHostEnvironment GetHostingEnvironment(this IServiceCollection services)
         {
+            if (!services.Any(d => d.ServiceType == typeof(IWebHostEnvironment)))
```

Comment:
This code searches the array 2 times (one in GetSingletonInstance). Use this instead:

`return services.GetSingletonInstanceOrNull<IWebHostEnvironment>() ?? new EmptyHostingEnvironment()...`;